Android support

crudelios opened the pull request:
Continuing from #195.

This PR adds support for Android.

**Building the apk**

First, copy the `SDL` and `SDL_mixer` sources to `julius/ext/SDL2`.

Then either:

* Download [Android Studio](https://developer.android.com/studio) and install the SDK as well as its NDK and CMake dependencies.

* Select `Open existing project` and select folder `android`.

Or:

* Install [Android Studio](https://developer.android.com/studio) for command line and run, from the julius directory, `cd android && ./gradlew assembleDebug`

    A docker image is available at https://hub.docker.com/r/npetrovsky/docker-android-sdk-ndk, preventing the need to download Android SDK. Just beware as the image itself is almost 2GB in size when compressed.

**Note:** If you already had Android Studio, make sure to update it, along with its plugins and tools, otherwise the build may fail.

In either case, the `apk` will be built to the short and easy to remember path `julius/android/distribution/android/julius/outputs/apk/debug`.

@bvschaik There are still many changes (about 1500 lines changed). However most of them are gradle scripts and configurations, the code changes should be about 750 lines.

@m-mohr, @411752230: As promised, this new build includes the improved touch support. Please test and report problems. For touch related bugs, please add them to #381, leaving this PR for specific Android related issues.

[**Download the latest apk here**](https://drive.google.com/open?id=1RaIQh5Y7H8bviobp7f84cieve0R7RAZ6) (latest version added on June 15th, 2020).

m-mohr commented:
Great! Could you provide an APK?

crudelios commented:
> Great! Could you provide an APK?

Lol I forgot to add it, sorry! Check the main comment again to get it.

crudelios commented on .ci_scripts/run_build.sh:
```diff
@@ -26,6 +26,10 @@ case "$BUILD_TARGET" in
 	;;
+"android")
+	cd android && TERM=dumb ./gradlew assembleDebug && \
+	cd .. && cp android/distribution/android/julius/outputs/apk/debug/julius-debug.apk build/julius.apk
```
I know. I haven't done it because I'm not sure how we should proceed with code signing.

daflame76 replied:
Not sure this is the best way, but: https://www.appveyor.com/docs/how-to/secure-files/ you can use that to secure a signing key.  It is only as secure as the appveyor account however.

This is an example with Travis-CI: https://android.jlelse.eu/using-travisci-to-securely-build-and-deploy-a-signed-version-of-your-android-app-94afdf5cf5b4

crudelios commented on android/julius/build.gradle:
```diff
@@ -0,0 +1,128 @@
+buildDir = file("../distribution/${rootProject.name}/${project.name}")
+
+task getVersion {
+    def project_version_major = 1
+    def project_version_minor = 3
+    def project_version_patch = 1
+    def is_release_version = false
```
I wish I knew how... but unless both CMake and gradle read from a common file, I don't know how to do it!

bvschaik replied:
I'll look into this. We should be able to run some code from the `build.gradle` file.

crudelios replied:
Thanks, I really appreciate your help on this one!

bvschaik replied:
This seems to work, reading the variables from the CMakeLists.txt file:
```suggestion
    def cmakeFile = rootProject.file("../CMakeLists.txt").text
    def project_version_major = (cmakeFile =~ /set\(PROJECT_VERSION_MAJOR (\d+)\)/)[0][1].toInteger()
    def project_version_minor = (cmakeFile =~ /set\(PROJECT_VERSION_MINOR (\d+)\)/)[0][1].toInteger()
    def project_version_patch = (cmakeFile =~ /set\(PROJECT_VERSION_PATCH (\d+)\)/)[0][1].toInteger()
    def is_release_version = (cmakeFile =~ /set\(IS_RELEASE_VERSION ([A-Z]+)\)/)[0][1].toBoolean()
```

crudelios commented on android/julius/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,35 @@
+            </intent-filter>
+        </activity>
+
+        <meta-data android:name="com.samsung.android.keepalive.density" android:value="true" />
```
If I remember well, it's needed to get the screen dpi

bvschaik replied:
It shouldn't. This property is Samsung-specific. I'd rather not have vendor specific code in the codebase.

crudelios replied:
Oh you're right, I remember now, I added this to be able to play Julius using Samsung DeX!

Julius is fully compatible with DeX and runs very well, so can we please keep it?

m-mohr replied:
DeX support would be great.

bvschaik replied:
At least add a comment explaining why we need this. What the heck is DeX?

crudelios replied:
I'll have at look if it still works fine without the line (it might just force close the app when switching in and out of DeX without that line), if it still works then the line can be safely removed.

crudelios replied:
DeX is a PC-like mode for android when connected to a monitor using HDMI or to a PC using the USB interface and the Samsung DeX app for Windows).

With DeX, you can play Caesar 3 on Android as you would on a regular PC:

![dex](https://user-images.githubusercontent.com/6518876/81160996-c4e3b800-8f82-11ea-80e3-b0160b13940e.JPG)

crudelios commented on src/platform/file_manager.c:
```diff
@@ -136,11 +140,12 @@ int platform_file_manager_list_directory_contents(const char *dir, int type, con
 int platform_file_manager_should_case_correct_file(void)
 {
-#ifdef _WIN32
+#if defined(_WIN32) || defined(__ANDROID__)
```
It is, but the case correctness is automatically handled on the Java side, so it's redundant to do it on the native side as well.

crudelios commented on android/julius/src/main/java/com/github/bvschaik/julius/FileManager.java:
```diff
@@ -0,0 +1,257 @@
+            if(baseUri == Uri.EMPTY) {
+                return null;
+            }
+            String[] filePart = filePath.split("[\\\\/]");
```
This was a while ago, but from what I remember this regex was the one that consistently worked with the hardcoded video paths in the language files.

I'll have a look at your suggestions and see if everything loads properly as well.

crudelios commented on android/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,6 @@
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip
```
Yes, but I think the Android Studio gradle plugin doesn't support it. At least from the official Android docs, 5.6.4 is the latest supported version!

bvschaik replied:
It does support it. I've been using 6.3 for 1.5 month at work, and we upgrade any time there's a new version available.

crudelios replied:
Thanks, I'll have a look then. Also, it seems [Gradle 6.4 has been released yesterday](https://gradle.org/releases/).

crudelios commented on cmake/FindSDL2.cmake:
```diff
@@ -65,71 +65,88 @@
+    SET(SDL2_ANDROID_HOOK ${SDL_EXT_DIR}/src/main/android/SDL_android_main.c)
+    ADD_LIBRARY(${SDL2_LIBRARY} SHARED IMPORTED)
+    SET_TARGET_PROPERTIES(${SDL2_LIBRARY} PROPERTIES IMPORTED_LOCATION
+    ${PROJECT_SOURCE_DIR}/android/distribution/android/SDL2/intermediates/ndkBuild/${ANDROID_BUILD_DIR}/obj/local/${ANDROID_ABI}/libSDL2.so)
```
It's the path where Gradle stores the built files, and `FIND_FILE` has never really worked that well on CMake for Android...

crudelios replied:
After having a look at this, I don't really know of a better way do integrate our existing code with Gradle. So this will have to stick :\

daflame76 replied:
It is weird to me too, but I also failed to find a way around it.  Strange Android build magic because it is merging native and Java code.  I think there is supposed to be a way it can be improved using mklink though to connect the SDL library to the julius android path.

crudelios replied:
Yeah, Android doesn't have much support for native libraries. It would also be great if we could store the pre built SDL libraries somewhere and reuse them, as it would drastically reduce the build times for travis.

pingwinator commented:
please, add mouse support

crudelios commented:
> please, add mouse support

There's already mouse support for android.

pingwinator commented:
Yes, but the right button doesn't work

Sent from nokia 3310 linux edition

> 6 мая 2020 г., в 20:23, José Cadete <notifications@github.com> написал(а):
> 
> ﻿
> please, add mouse support
> 
> There's already mouse support for android.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

crudelios commented:
> Yes, but the right button doesn't work

Hummm... it doesn't show the building info. I'll have a look at that!

crudelios commented:
Fixing the right mouse click is harder than I expected as all Android devices map right mouse button the back button.

Therefore, I won't fix it for now.

crudelios commented on src/platform/android/android.c:
```diff
@@ -0,0 +1,216 @@
+    int len = (*handler.env)->GetArrayLength(handler.env, result);
+    for (int i = 0; i < len; ++i) {
+        jstring jfilename = (jstring) (*handler.env)->GetObjectArrayElement(handler.env, result, i);
+        const char *filename = (*handler.env)->GetStringUTFChars(handler.env, jfilename, NULL);
```
All `GetStringUTFChars` are memleaking as they must be released by calling `ReleaseStringUTFChars`, so I must fix this before merge.

crudelios commented:
@pingwinator can you please have a look at the latest version? (You can download it at the top). I added a hack to allow right click input, it's working on my Samsung Galaxy S10 but I don't know if it works on every Android device.

Thanks.

pingwinator commented:
@crudelios Hi.
I tested it on pixel 3a and Samsung tab s5e. It works perfectly on the pixel and Samsung in DeX mode. But the right button doesn't work in the tablet mode on s5e.
Thanks for the fast support!

crudelios commented:
@pingwinator Hey, thanks for the feedback. I'm guessing that on Samsung Tab S5E, right-clicking on the city still opens the `Exit game` window, right?

I'm not sure if I'll be able to fix this, as the problem lies within the Android environment itself and the subsequent SDL's input handling.

I already implemented a very hackish solution (`Esc` keypresses are being interpreted as mouse right-clicks when on Android using mouse with DeX disabled) and don't really want to add any further hacks. In fact, even this hack I added may not make it to the merging of this PR.

pingwinator commented:
> Hey, thanks for the feedback. I'm guessing that on Samsung Tab S5E, right-clicking on the city still opens the Exit game window, right?

nope, nothing happened.

> I'm not sure if I'll be able to fix this, as the problem lies within the Android environment itself and the subsequent SDL's input handling.

> I already implemented a very hackish solution (Esc keypresses are being interpreted as mouse right-clicks when on Android using mouse with DeX disabled) and don't really want to add any further hacks. In fact, even this hack I added may not make it to the merging of this PR.

yes, I agree. This is a huge area for future improvements. Let's merge it.

crudelios commented:
@pingwinator I have made a slight change to the hack that hopefully will detect the right mouse click in your Galaxy Tab SSE.

Grab the test apk here: https://drive.google.com/open?id=17h2mheXGofBeN7DEOSqeSyrrb7TaOumy

Can you please test and report? I have no way to test this.

pingwinator commented:
HI.
No, no changes.

> On 13. May 2020, at 00:39, José Cadete <notifications@github.com> wrote:
> 
> 
> @pingwinator I have made a slight change to the hack that hopefully will detect the right mouse click in your Galaxy Tab SSE.
> 
> Grab the test apk here: https://drive.google.com/open?id=17h2mheXGofBeN7DEOSqeSyrrb7TaOumy
> 
> Can you please test and report? I have no way to test this.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

bvschaik commented:
Closing this PR since the code is part of #514 which includes the required configuration for release builds.
